accounts/abi/bind/dep_tree_test.go

```diff
@@ -0,0 +1,309 @@
+)
+
+type linkTestCase struct {
+	// map of pattern to unlinked bytecode (for the purposes of tests just contains the patterns of its dependencies)
+	libCodes      map[string]string
+	contractCodes map[string]string
+
+	overrides map[string]common.Address
+}
+
+func makeLinkTestCase(input map[rune][]rune, overrides map[rune]common.Address) *linkTestCase {
+	codes := make(map[string]string)
+	libCodes := make(map[string]string)
+	contractCodes := make(map[string]string)
+
+	inputMap := make(map[rune]map[rune]struct{})
+	// set of solidity patterns for all contracts that are known to be libraries
+	libs := make(map[string]struct{})
+
+	// map of test contract id (rune) to the solidity library pattern (hash of that rune)
+	patternMap := map[rune]string{}
+
+	for contract, deps := range input {
+		inputMap[contract] = make(map[rune]struct{})
+		if _, ok := patternMap[contract]; !ok {
+			patternMap[contract] = crypto.Keccak256Hash([]byte(string(contract))).String()[2:36]
+		}
+
+		for _, dep := range deps {
+			if _, ok := patternMap[dep]; !ok {
+				patternMap[dep] = crypto.Keccak256Hash([]byte(string(dep))).String()[2:36]
+			}
+			codes[patternMap[contract]] = codes[patternMap[contract]] + fmt.Sprintf("__$%s$__", patternMap[dep])
+			inputMap[contract][dep] = struct{}{}
+			libs[patternMap[dep]] = struct{}{}
+		}
+	}
+	overridesPatterns := make(map[string]common.Address)
+	for contractId, overrideAddr := range overrides {
+		pattern := crypto.Keccak256Hash([]byte(string(contractId))).String()[2:36]
+		overridesPatterns[pattern] = overrideAddr
+	}
+
+	for _, pattern := range patternMap {
+		if _, ok := libs[pattern]; ok {
+			// if the library didn't depend on others, give it some dummy code to not bork deployment logic down-the-line
+			if len(codes[pattern]) == 0 {
+				libCodes[pattern] = "ff"
+			} else {
+				libCodes[pattern] = codes[pattern]
+			}
+		} else {
+			contractCodes[pattern] = codes[pattern]
+		}
+	}
+
+	return &linkTestCase{
+		libCodes,
+		contractCodes,
+		overridesPatterns,
+	}
+}
+
+var testKey, _ = crypto.HexToECDSA("b71c71a67e1177ad4e901695e1b4b9ee17ae16c6668d313eac2f96dbcda3f291")
+
+type linkTestCaseInput struct {
+	input          map[rune][]rune
+	overrides      map[rune]struct{}
+	expectDeployed map[rune]struct{}
+}
+
+// linkDeps will return a set of root dependencies and their sub-dependencies connected via the Deps field
+func linkDeps(deps map[string]*MetaData) []*MetaData {
+	roots := make(map[string]struct{})
+	for pattern, _ := range deps {
+		roots[pattern] = struct{}{}
+	}
+
+	connectedDeps := make(map[string]MetaData)
+	for pattern, dep := range deps {
+		connectedDeps[pattern] = __linkDeps(*dep, deps, &roots)
+	}
+	rootMetadatas := []*MetaData{}
+	for pattern, _ := range roots {
+		dep := connectedDeps[pattern]
+		rootMetadatas = append(rootMetadatas, &dep)
+	}
+	return rootMetadatas
+}
+
+func __linkDeps(metadata MetaData, depMap map[string]*MetaData, roots *map[string]struct{}) MetaData {
```

Comment:
What's up with this name? Why such a strange prefix?

Reply:
that's the recursing part of `linkDeps`.  I didn't know what to call it.  That code needs a bit of love (documentation/renaming where appropriate) but I'm trying to port the v1 bind tests over to v2 atm.

Reply:
Well, that's now how we name things in go. I'm sure you can figure out a more descriptive name